File: apps/api/app/Services/TimeAccountingService.php
<?php

namespace App\Services;

use Carbon\CarbonImmutable;
use Illuminate\Support\Collection;

final class TimeAccountingService
{
    /**
     * @param Collection<int, object> $sessions Objects exposing started_at and ended_at.
     * @return array{effort_seconds:int, elapsed_coverage_seconds:int, concurrent_effort_seconds:int}
     */
    public function summarize(Collection $sessions, CarbonImmutable $rangeStart, CarbonImmutable $rangeEnd): array
    {
        $effort = 0;
        $intervals = [];

        foreach ($sessions as $session) {
            $startedAt = CarbonImmutable::parse($session->started_at);
            $endedAt = CarbonImmutable::parse($session->ended_at);

            $start = $startedAt->greaterThan($rangeStart) ? $startedAt : $rangeStart;
            $end = $endedAt->lessThan($rangeEnd) ? $endedAt : $rangeEnd;

            if ($end->lessThanOrEqualTo($start)) {
                continue;
            }

            $seconds = $start->diffInSeconds($end);
            $effort += $seconds;
            $intervals[] = [$start, $end];
        }

        $coverage = $this->unionSeconds($intervals);

        return [
            'effort_seconds' => $effort,
            'elapsed_coverage_seconds' => $coverage,
            'concurrent_effort_seconds' => max(0, $effort - $coverage),
        ];
    }

    /** @param array<int, array{0:CarbonImmutable,1:CarbonImmutable}> $intervals */
    private function unionSeconds(array $intervals): int
    {
        if ($intervals === []) {
            return 0;
        }

        usort($intervals, static fn (array $a, array $b): int => $a[0]->getTimestamp() <=> $b[0]->getTimestamp());

        [$currentStart, $currentEnd] = $intervals[0];
        $total = 0;

        foreach (array_slice($intervals, 1) as [$start, $end]) {
            if ($start->lessThanOrEqualTo($currentEnd)) {
                if ($end->greaterThan($currentEnd)) {
                    $currentEnd = $end;
                }
                continue;
            }

            $total += $currentStart->diffInSeconds($currentEnd);
            $currentStart = $start;
            $currentEnd = $end;
        }

        return $total + $currentStart->diffInSeconds($currentEnd);
    }
}

File: apps/windows-agent/WorkTracker.Agent/Services/TimeAccountingService.cs
using WorkTracker.Agent.Domain;

namespace WorkTracker.Agent.Services;

public sealed record TimeAccountingSummary(int EffortSeconds, int ElapsedCoverageSeconds, int ConcurrentEffortSeconds);

public static class TimeAccountingService
{
    public static TimeAccountingSummary Summarize(IEnumerable<ActivitySession> sessions, DateTimeOffset rangeStart, DateTimeOffset rangeEnd, int additionalEffortSeconds = 0)
    {
        var intervals = new List<(DateTimeOffset Start, DateTimeOffset End)>();
        var effort = Math.Max(0, additionalEffortSeconds);

        foreach (var session in sessions)
        {
            var start = session.StartedAt < rangeStart ? rangeStart : session.StartedAt;
            var end = session.EndedAt > rangeEnd ? rangeEnd : session.EndedAt;
            if (end <= start) continue;

            effort += Math.Max(0, (int)Math.Floor((end - start).TotalSeconds));
            intervals.Add((start, end));
        }

        var coverage = UnionSeconds(intervals);
        return new TimeAccountingSummary(effort, coverage, Math.Max(0, effort - coverage));
    }

    private static int UnionSeconds(List<(DateTimeOffset Start, DateTimeOffset End)> intervals)
    {
        if (intervals.Count == 0) return 0;
        intervals.Sort((a, b) => a.Start.CompareTo(b.Start));

        var total = 0d;
        var currentStart = intervals[0].Start;
        var currentEnd = intervals[0].End;

        for (var i = 1; i < intervals.Count; i++)
        {
            var next = intervals[i];
            if (next.Start <= currentEnd)
            {
                if (next.End > currentEnd) currentEnd = next.End;
                continue;
            }

            total += (currentEnd - currentStart).TotalSeconds;
            currentStart = next.Start;
            currentEnd = next.End;
        }

        total += (currentEnd - currentStart).TotalSeconds;
        return Math.Max(0, (int)Math.Floor(total));
    }
}
